src/Type/Php/SubstrDynamicReturnTypeExtension.php

```diff
@@ -40,13 +46,43 @@ public function getTypeFromFunctionCall(
 
 			$negativeOffset = IntegerRangeType::fromInterval(null, -1)->isSuperTypeOf($offset)->yes();
 			$zeroOffset = (new ConstantIntegerType(0))->isSuperTypeOf($offset)->yes();
+			$length = null;
 			$positiveLength = false;
 
 			if (count($args) === 3) {
 				$length = $scope->getType($args[2]->value);
 				$positiveLength = IntegerRangeType::fromInterval(1, null)->isSuperTypeOf($length)->yes();
 			}
 
+			$constantStrings = TypeUtils::getConstantStrings($string);
+			if (count($constantStrings) > 0 &&
+				$offset instanceof ConstantIntegerType &&
+				($length === null || $length instanceof ConstantIntegerType)) {
+				$results = [];
+				foreach ($constantStrings as $constantString) {
+					if ($length !== null) {
+						$substr = substr(
+							$constantString->getValue(),
+							$offset->getValue(),
+							$length->getValue(),
+						);
+					} else {
+						$substr = substr(
+							$constantString->getValue(),
+							$offset->getValue(),
+						);
+					}
+
+					if (is_bool($substr)) {
+						$results[] = new ConstantBooleanType($substr);
+					} else {
+						$results[] = new ConstantStringType($substr);
+					}
```

Comment:
You could use `ConstantTypeHelper::getTypeFromValue()` instead (even though it's deprecated)

Reply:
I searched the code-base yesterday for this method .. I knew we had such thing, but wasn't able to find it :)

as you mentioned - the method is deprecated and seems to be a rather big canon for the rather small problem I solve here - I feel the PR should stay as is.